File: python/cod4demo/servercmd.py
"""Decoding of reliable server commands (``svc_serverCommand`` strings).

The first character of a command selects the handler (see
``CG_DeployServerCommand`` in cg_servercmds_mp.cpp and
``CL_CGameNeedsServerCommand`` in cl_cgame_mp.cpp). ``decode()`` returns a
dict with the verb, a readable name and the decoded arguments.
"""

from __future__ import annotations

from .constants import SERVER_COMMANDS
from .text import strip_colors, tokenize

#: handled by the client engine itself before cgame sees them
ENGINE_COMMANDS = {
    "w": "disconnect",                 # w <reason> [PB]
    "x": "big_configstring_start",     # x <index> <part>
    "y": "big_configstring_append",    # y <index> <part>
    "z": "big_configstring_end",       # z <index> <part>
}


def _int(s: str, default: int | None = None) -> int | None:
    try:
        return int(s)
    except (TypeError, ValueError):
        try:
            return int(float(s))
        except (TypeError, ValueError):
            return default


def _float(s: str) -> float | None:
    try:
        return float(s)
    except (TypeError, ValueError):
        return None


def decode(text: str) -> dict:
    """Decode one server command string."""
    verb = text[:1] if text else ""
    name = ENGINE_COMMANDS.get(verb) or SERVER_COMMANDS.get(verb) or "unknown"
    out: dict = {"verb": verb, "name": name}
    if verb in ("d", "x", "y", "z"):
        argv = tokenize(text, 3)
        out["index"] = _int(argv[1]) if len(argv) > 1 else None
        out["value"] = argv[2] if len(argv) > 2 else ""
        return out
    argv = tokenize(text)
    args = argv[1:]
    out["args"] = args
    if verb == "b" and args:
        n = _int(args[0], 0) or 0
        out["count"] = n
        out["score_axis"] = _int(args[1]) if len(args) > 1 else None
        out["score_allies"] = _int(args[2]) if len(args) > 2 else None
        out["scorelimit"] = _int(args[3]) if len(args) > 3 else None
        entries = []
        for i in range(min(n, 64)):
            base = 4 + 7 * i
            if base + 7 > len(args):
                break
            c, score, ping, deaths, icon, kills, assists = (_int(a) for a in args[base:base + 7])
            entries.append({"client": c, "score": score, "ping": ping, "deaths": deaths,
                            "status_icon": icon, "kills": kills, "assists": assists})
        out["entries"] = entries
    elif verb in ("h", "i"):
        out["text"] = args[0] if args else ""
        out["scope"] = "all" if verb == "h" else "team"
    elif verb in ("c", "e", "f", "g"):
        out["text"] = args[0] if args else ""
    elif verb in ("G", "H"):
        out["team"] = "axis" if verb == "G" else "allies"
        out["score"] = _int(args[0]) if args else None
    elif verb == "I" and len(args) >= 2:
        out["client"] = _int(args[0])
        out["score"] = _int(args[1])
    elif verb in ("a", "C") and args:
        out["weapon"] = _int(args[0])
    elif verb == "J" and args:
        out["menu"] = _int(args[0])
    elif verb == "K" and args:
        out["client"] = _int(args[0])
    elif verb == "N" and len(args) >= 2:
        out["stat"] = _int(args[0])
        out["value"] = _int(args[1])
    elif verb == "v":
        out["dvars"] = [(args[i], args[i + 1] if i + 1 < len(args) else "")
                        for i in range(0, len(args), 2)]
    elif verb == "t" and args:
        out["menu"] = _int(args[0])
    elif verb == "o" and args:
        out["alias"] = args[0]
        out["volume"] = _int(args[1]) if len(args) > 1 else None
    elif verb == "p" and args:
        out["fade_time"] = _int(args[0])
    elif verb == "q" and len(args) >= 2:
        out["volume"] = _float(args[0])
        out["time"] = _int(args[1])
    elif verb == "s" and args:
        out["alias"] = args[0]
    elif verb == "w":
        out["reason"] = args[0] if args else ""
    elif verb == "j" and len(args) >= 8:
        out["id"] = _int(args[0])
        out["draw_type"] = _int(args[1])
        out["pos"] = [_float(a) for a in args[2:5]]
        out["dir"] = [_float(a) for a in args[5:8]]
    return out


class BigConfigStringAssembler:
    """Reassembles ``x`` / ``y`` / ``z`` pieces into one config string update."""

    def __init__(self) -> None:
        self.index: int | None = None
        self.buf = ""

    def feed(self, d: dict) -> tuple[int, str] | None:
        verb = d["verb"]
        if verb == "x":
            self.index = d["index"]
            self.buf = d["value"]
        elif verb == "y":
            self.buf += d["value"]
        elif verb == "z":
            self.buf += d["value"]
            if self.index is not None:
                done = (self.index, self.buf)
                self.index, self.buf = None, ""
                return done
        return None


def clean(text: str) -> str:
    return strip_colors(text)

File: js/common/servercmd.js
/* Decoding of reliable server commands (first character selects the handler,
 * CG_DeployServerCommand / CL_CGameNeedsServerCommand). */
C4.define('servercmd', function (C4) {
  'use strict';
  const { tokenize } = C4.text;
  const NAMES = {
    B: 'map_restart', C: 'set_equipped_offhand', D: 'reverb_deactivate', E: 'channel_volume_set',
    F: 'channel_volume_deactivate', G: 'team_score_axis', H: 'team_score_allies', I: 'client_score',
    J: 'menu_show_notify', K: 'reset_player_muting', L: 'close_ingame_menu', N: 'set_stat',
    a: 'select_weapon', b: 'scoreboard', c: 'announcement', d: 'configstring', e: 'print',
    f: 'game_message', g: 'bold_game_message', h: 'chat', i: 'team_chat', j: 'dynent_destroy',
    k: 'local_sound_stop', n: 'map_restart_persist', o: 'play_music', p: 'stop_music',
    q: 'fade_sounds', r: 'reverb', s: 'local_sound', t: 'open_script_menu', u: 'close_script_menu',
    v: 'set_client_dvars', w: 'disconnect', x: 'big_configstring_start',
    y: 'big_configstring_append', z: 'big_configstring_end'
  };
  const int = s => { const v = parseInt(s, 10); return Number.isFinite(v) ? v : null; };

  function decode(text) {
    const verb = text ? text[0] : '';
    const out = { verb, name: NAMES[verb] || 'unknown' };
    if (verb === 'd' || verb === 'x' || verb === 'y' || verb === 'z') {
      const argv = tokenize(text, 3);
      out.index = argv.length > 1 ? int(argv[1]) : null;
      out.value = argv.length > 2 ? argv[2] : '';
      return out;
    }
    const args = tokenize(text).slice(1);
    out.args = args;
    switch (verb) {
      case 'b': {
        const n = int(args[0]) || 0;
        out.scoreAxis = int(args[1]);
        out.scoreAllies = int(args[2]);
        out.scorelimit = int(args[3]);
        out.entries = [];
        for (let i = 0; i < Math.min(n, 64); i++) {
          const b = 4 + 7 * i;
          if (b + 7 > args.length) break;
          const v = args.slice(b, b + 7).map(int);
          out.entries.push({ client: v[0], score: v[1], ping: v[2], deaths: v[3], statusIcon: v[4], kills: v[5], assists: v[6] });
        }
        break;
      }
      case 'h': case 'i': out.text = args[0] || ''; out.scope = verb === 'h' ? 'all' : 'team'; break;
      case 'c': case 'e': case 'f': case 'g': out.text = args[0] || ''; break;
      case 'G': case 'H': out.team = verb === 'G' ? 'axis' : 'allies'; out.score = int(args[0]); break;
      case 'I': out.client = int(args[0]); out.score = int(args[1]); break;
      case 'v':
        out.dvars = [];
        for (let i = 0; i < args.length; i += 2) out.dvars.push([args[i], args[i + 1] == null ? '' : args[i + 1]]);
        break;
      case 'w': out.reason = args[0] || ''; break;
      default: break;
    }
    return out;
  }

  /** Reassembles x / y / z pieces into one config string update. */
  class BigConfigString {
    constructor() { this.index = null; this.buf = ''; }
    feed(d) {
      if (d.verb === 'x') { this.index = d.index; this.buf = d.value; }
      else if (d.verb === 'y') this.buf += d.value;
      else if (d.verb === 'z') {
        this.buf += d.value;
        if (this.index != null) { const r = [this.index, this.buf]; this.index = null; this.buf = ''; return r; }
      }
      return null;
    }
  }

  C4.servercmd = { decode, BigConfigString, NAMES };
});
